libraries/AP_Compass/AP_Compass_MMC3416.cpp

```diff
@@ -0,0 +1,338 @@
+                                              enum Rotation rotation)
+{
+    if (!dev) {
+        return nullptr;
+    }
+    AP_Compass_MMC3416 *sensor = new AP_Compass_MMC3416(compass, std::move(dev), force_external, rotation);
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_MMC3416::AP_Compass_MMC3416(Compass &compass,
+                                       AP_HAL::OwnPtr<AP_HAL::Device> _dev,
+                                       bool _force_external,
+                                       enum Rotation _rotation)
+    : AP_Compass_Backend(compass)
+    , dev(std::move(_dev))
+    , force_external(_force_external)
+    , rotation(_rotation)
+{
+}
+
+bool AP_Compass_MMC3416::init()
+{
+    if (!dev->get_semaphore()->take(0)) {
+        return false;
+    }
+
+    dev->set_retries(10);
+    
+    uint8_t whoami;
+    if (!dev->read_registers(REG_PRODUCT_ID, &whoami, 1) ||
+        whoami != 0x06) {
+        // not a MMC3416
+        goto fail;
+    }
+
+    // reset sensor
+    dev->write_register(REG_CONTROL1, 0x80);
+    hal.scheduler->delay(10);
+    
+    dev->write_register(REG_CONTROL0, 0x00); // single shot
+    dev->write_register(REG_CONTROL1, 0x00); // 16 bit, 7.92ms
+    
+    dev->get_semaphore()->give();
+
+    /* register the compass instance in the frontend */
+    compass_instance = register_compass();
+
+    printf("Found a MMC3416 on 0x%x as compass %u\n", dev->get_bus_id(), compass_instance);
+    
+    set_rotation(compass_instance, rotation);
+
+    if (force_external) {
+        set_external(compass_instance, true);
+    }
+    
+    dev->set_device_type(DEVTYPE_MMC3416);
+    set_dev_id(compass_instance, dev->get_bus_id());
+
+    dev->set_retries(1);
+    
+    // call timer() at 100Hz
+    dev->register_periodic_callback(10000,
+                                    FUNCTOR_BIND_MEMBER(&AP_Compass_MMC3416::timer, bool));
+
+    // wait 250ms for the compass to make it's initial readings
+    hal.scheduler->delay(250);
+    
+    return true;
+
+fail:
+    dev->get_semaphore()->give();
+    return false;
+}
+
+bool AP_Compass_MMC3416::timer()
+{
+    const uint16_t measure_count_limit = 50;
+    const uint16_t zero_offset = 32768; // 16 bit mode
+    const uint16_t sensitivity = 2048; // counts per Gauss, 16 bit mode
+    const float counts_to_milliGauss = 1.0e3f / sensitivity;
+
+    uint32_t now = AP_HAL::millis();
+    if (now - last_sample_ms > 500) {
+        // seems to be stuck or on first sample, reset state machine
+        state = STATE_REFILL1;
+        last_sample_ms = now;
+    }
+    
+    /*
+      we use the SET/RESET method to remove bridge offset every
+      measure_count_limit measurements. This involves a fairly complex
+      state machine, but means we are much less sensitive to
+      temperature changes
+     */
+    switch (state) {
+    case STATE_REFILL1:
```

Comment:
The description of the switch statement is new. I think that it is better to make it a new description.